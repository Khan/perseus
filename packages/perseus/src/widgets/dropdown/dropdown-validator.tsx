import type {PerseusScore} from "../../types";
import type {
    PerseusDropdownRubric,
    PerseusDropdownUserInput,
} from "../../validation.types";

function dropdownValidator(
    userInput: PerseusDropdownUserInput,
    rubric: PerseusDropdownRubric,
): PerseusScore {
    const selected = userInput.value;
    if (selected === 0) {
        return {
            type: "invalid",
            message: null,
        };
    }
    const correct = rubric.choices[selected - 1].correct;
    return {
        type: "points",
        earned: correct ? 1 : 0,
        total: 1,
        message: null,
    };
}

export default dropdownValidator;
