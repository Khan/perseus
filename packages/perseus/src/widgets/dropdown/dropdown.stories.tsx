import * as React from "react";

import {RendererWithDebugUI} from "../../../../../testing/renderer-with-debug-ui";
import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";
import {generateTestPerseusItem} from "../../util/test-utils";

import {
    basicDropdown,
    dropdownWithEmptyPlaceholder,
    dropdownWithMath,
    dropdownWithVisibleLabel,
    inlineDropdownWithVisibleLabel,
} from "./dropdown.testdata";

export default {
    title: "Perseus/Widgets/Dropdown",
};

type StoryArgs = Record<any, any>;

export const BasicDropdown = (args: StoryArgs): React.ReactElement => {
    return <RendererWithDebugUI question={basicDropdown} />;
};

export const DropdownWithMath = (args: StoryArgs): React.ReactElement => {
    return <RendererWithDebugUI question={dropdownWithMath} />;
};

export const DropdownWithVisibleLabel = (
    args: StoryArgs,
): React.ReactElement => {
    return <RendererWithDebugUI question={dropdownWithVisibleLabel} />;
};

export const InlineDropdownWithVisibleLabel = (
    args: StoryArgs,
): React.ReactElement => {
    return <RendererWithDebugUI question={inlineDropdownWithVisibleLabel} />;
};

export const DropdownWithEmptyPlaceholder = (
    args: StoryArgs,
): React.ReactElement => {
    return <RendererWithDebugUI question={dropdownWithEmptyPlaceholder} />;
};

export const AnswerlessBasicDropdown = (
    args: StoryArgs,
): React.ReactElement => {
    return (
        <ServerItemRendererWithDebugUI
            item={generateTestPerseusItem({
                question: basicDropdown,
            })}
            startAnswerless={true}
        />
    );
};
