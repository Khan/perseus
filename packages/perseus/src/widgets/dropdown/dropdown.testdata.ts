import type {PerseusRenderer} from "../../perseus-types";

export const question1: PerseusRenderer = {
    content:
        "The total number of boxes the forklift can carry is [[☃ dropdown 1]] $60$.",
    images: {},
    widgets: {
        "dropdown 1": {
            type: "dropdown",
            alignment: "default",
            static: false,
            graded: true,
            options: {
                static: false,
                placeholder: "greater/less than or equal to",
                choices: [
                    {
                        content: "greater than or equal to",
                        correct: false,
                    },
                    {
                        content: "less than or equal to",
                        correct: true,
                    },
                ],
            },
            version: {
                major: 0,
                minor: 0,
            },
        },
    },
};

export const dropdownWithVisibleLabel: PerseusRenderer = {
    content: "[[☃ dropdown 1]]",
    images: {},
    widgets: {
        "dropdown 1": {
            type: "dropdown",
            alignment: "default",
            static: false,
            graded: true,
            options: {
                static: false,
                placeholder: "Choose an answer",
                choices: [
                    {
                        content: "True",
                        correct: true,
                    },
                    {
                        content: "False",
                        correct: false,
                    },
                ],
                visibleLabel: "Test label",
                ariaLabel: "Test ARIA label",
            },
            version: {
                major: 0,
                minor: 0,
            },
        },
        "dropdown 2": {
            type: "dropdown",
            alignment: "default",
            static: false,
            graded: true,
            options: {
                static: false,
                placeholder: "",
                choices: [
                    {
                        content: "True",
                        correct: true,
                    },
                    {
                        content: "False",
                        correct: false,
                    },
                ],
                visibleLabel: "Test label",
                ariaLabel: "Test ARIA label",
            },
            version: {
                major: 0,
                minor: 0,
            },
        },
    },
};

export const inlineDropdownWithVisibleLabel: PerseusRenderer = {
    content:
        "The dropdown widget is often used inline. This is how it would look in an article with the new visible label:\n\nLorem ipsum odor amet, consectetuer adipiscing elit. Mus curae sollicitudin penatibus, mattis suscipit habitant tincidunt mauris. Vitae curae dolor gravida vehicula adipiscing vulputate penatibus. [[☃ dropdown 1]] Ultricies mollis taciti vel, penatibus dapibus interdum pharetra. Ultricies sollicitudin facilisi vehicula dapibus ligula maecenas libero ligula. Lobortis luctus accumsan rhoncus posuere sapien mi habitant fusce. Per ultrices ac mus ligula habitant pulvinar aliquam dui lacus.\n\nAnother use case is that it can be used in tables:\n\nheader 1 | header 2 \n- | -\ndata 1 | [[☃ dropdown 2]]\ndata 4 | data 5\ndata 7 | data 8",
    images: {},
    widgets: {
        "dropdown 1": {
            type: "dropdown",
            alignment: "default",
            static: false,
            graded: true,
            options: {
                static: false,
                placeholder: "Choose an answer",
                choices: [
                    {
                        content: "True",
                        correct: true,
                    },
                    {
                        content: "False",
                        correct: false,
                    },
                ],
                visibleLabel: "Test label",
                ariaLabel: "Test ARIA label",
            },
            version: {
                major: 0,
                minor: 0,
            },
        },
        "dropdown 2": {
            type: "dropdown",
            alignment: "default",
            static: false,
            graded: true,
            options: {
                static: false,
                placeholder: "",
                choices: [
                    {
                        content: "True",
                        correct: true,
                    },
                    {
                        content: "False",
                        correct: false,
                    },
                ],
                visibleLabel: "Test label",
                ariaLabel: "Test ARIA label",
            },
            version: {
                major: 0,
                minor: 0,
            },
        },
    },
};
