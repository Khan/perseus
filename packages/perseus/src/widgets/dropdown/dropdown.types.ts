import type {PerseusDropdownWidgetOptions} from "../../perseus-types";

export type Rubric = PerseusDropdownWidgetOptions;

export type UserInput = {
    value: number;
};
