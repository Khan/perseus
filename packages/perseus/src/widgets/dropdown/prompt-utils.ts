import {WidgetType} from "../../prompt-types";

import type dropdown from "./dropdown";
import type {PerseusDropdownUserInput} from "../../validation.types";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type WidgetProps = PropsFor<typeof dropdown.widget>;

export type DropdownPromptJSON = {
    type: WidgetType;
    options: {
        items: WidgetProps["choices"];
    };
    userInput: {
        selectedIndex: PerseusDropdownUserInput["value"];
    };
};

export const getPromptJSON = (
    renderProps: WidgetProps,
    userInput: PerseusDropdownUserInput,
): DropdownPromptJSON => {
    return {
        type: WidgetType.DROPDOWN,
        options: {
            items: renderProps.choices,
        },
        userInput: {
            selectedIndex: userInput.value - 1, // Offset to account for placeholder
        },
    };
};
