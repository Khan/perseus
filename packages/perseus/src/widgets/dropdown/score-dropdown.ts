import validateDropdown from "./validate-dropdown";

import type {PerseusScore} from "../../types";
import type {
    PerseusDropdownScoringData,
    PerseusDropdownUserInput,
} from "../../validation.types";

function scoreDropdown(
    userInput: PerseusDropdownUserInput,
    scoringData: PerseusDropdownScoringData,
): PerseusScore {
    const validationError = validateDropdown(userInput);
    if (validationError) {
        return validationError;
    }
    const correct = scoringData.choices[userInput.value - 1].correct;
    return {
        type: "points",
        earned: correct ? 1 : 0,
        total: 1,
        message: null,
    };
}

export default scoreDropdown;
