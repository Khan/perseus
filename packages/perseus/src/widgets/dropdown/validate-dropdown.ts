import type {ValidationResult} from "../../types";
import type {PerseusDropdownUserInput} from "../../validation.types";

/**
 * Checks if the user has selected an item from the dropdown before scoring.
 * This is shown with a userInput value / index other than 0.
 */
function validateDropdown(
    userInput: PerseusDropdownUserInput,
): ValidationResult {
    if (userInput.value === 0) {
        return {
            type: "invalid",
            message: null,
        };
    }
    return null;
}

export default validateDropdown;
