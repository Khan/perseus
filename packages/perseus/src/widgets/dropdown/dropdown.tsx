import {UniqueIDProvider, View} from "@khanacademy/wonder-blocks-core";
import {SingleSelect, OptionItem} from "@khanacademy/wonder-blocks-dropdown";
import {LabelLarge} from "@khanacademy/wonder-blocks-typography";
import * as React from "react";
import ReactDOM from "react-dom";

import {PerseusI18nContext} from "../../components/i18n-context";
import {ApiOptions} from "../../perseus-api";
import {getPromptJSON as _getPromptJSON} from "../../widget-ai-utils/dropdown/dropdown-ai-utils";

import scoreDropdown from "./score-dropdown";

import type {PerseusDropdownWidgetOptions} from "../../perseus-types";
import type {Widget, WidgetExports, WidgetProps} from "../../types";
import type {
    PerseusDropdownRubric,
    PerseusDropdownUserInput,
} from "../../validation.types";
import type {DropdownPromptJSON} from "../../widget-ai-utils/dropdown/dropdown-ai-utils";

type Props = WidgetProps<RenderProps, PerseusDropdownRubric> & {
    selected: number;
};

type DefaultProps = {
    choices: Props["choices"];
    selected: Props["selected"];
    placeholder: Props["placeholder"];
    apiOptions: Props["apiOptions"];
};

class Dropdown extends React.Component<Props> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    static defaultProps: DefaultProps = {
        choices: [],
        selected: 0,
        placeholder: "",
        apiOptions: ApiOptions.defaults,
    };

    focus: () => boolean = () => {
        // TODO(LP-10797): This focus() call doesn't do anything because our
        // root element is a <div> and that cannot be focused without a
        // tabIndex.
        // @ts-expect-error - TS2531 - Object is possibly 'null'. | TS2339 - Property 'focus' does not exist on type 'Element | Text'.
        ReactDOM.findDOMNode(this).focus();
        return true;
    };

    _handleChangeEvent: (arg1: React.ChangeEvent<HTMLInputElement>) => void = (
        e,
    ) => {
        this._handleChange(parseInt(e.target.value));
    };

    _handleChange: (arg1: number) => void = (selected) => {
        this.props.trackInteraction();
        this.props.onChange({selected: selected});
    };

    getUserInput(): PerseusDropdownUserInput {
        return {value: this.props.selected};
    }

    getPromptJSON(): DropdownPromptJSON {
        return _getPromptJSON(this.props, this.getUserInput());
    }

    render(): React.ReactNode {
        const children = [
            <OptionItem
                key="placeholder"
                value="0"
                disabled
                label={this.props.placeholder}
            />,
            ...this.props.choices.map((choice, i) => (
                <OptionItem
                    key={String(i + 1)}
                    value={String(i + 1)}
                    label={choice}
                />
            )),
        ];

        return (
            <UniqueIDProvider scope="dropdown-widget" mockOnFirstRender={true}>
                {(ids) => (
                    <View
                        // NOTE(jared): These are required to prevent weird behavior
                        // When there's a dropdown in a zoomable table.
                        onClick={(e) => {
                            e.stopPropagation();
                        }}
                        onTouchStart={(e) => {
                            e.stopPropagation();
                        }}
                    >
                        {this.props.visibleLabel && (
                            <LabelLarge
                                tag="label"
                                id={ids.get("dropdown-label")}
                            >
                                {this.props.visibleLabel}
                            </LabelLarge>
                        )}
                        <SingleSelect
                            id={ids.get("dropdown")}
                            placeholder=""
                            onChange={(value) =>
                                this._handleChange(parseInt(value))
                            }
                            selectedValue={String(this.props.selected)}
                            disabled={this.props.apiOptions.readOnly}
                            aria-label={
                                this.props.ariaLabel ||
                                this.context.strings.selectAnAnswer
                            }
                            aria-labelledby={ids.get("dropdown-label")}
                            // This is currently necessary for SRs to read the labels properly.
                            // However, WB is working on a change to add the "combobox" role to
                            // all dropdowns.
                            // See https://khanacademy.atlassian.net/browse/WB-1671
                            role="combobox"
                        >
                            {children}
                        </SingleSelect>
                    </View>
                )}
            </UniqueIDProvider>
        );
    }
}

type RenderProps = {
    placeholder: PerseusDropdownWidgetOptions["placeholder"];
    visibleLabel: PerseusDropdownWidgetOptions["visibleLabel"];
    ariaLabel: PerseusDropdownWidgetOptions["ariaLabel"];
    choices: ReadonlyArray<string>;
};

const optionsTransform: (arg1: PerseusDropdownWidgetOptions) => RenderProps = (
    widgetOptions,
) => {
    return {
        placeholder: widgetOptions.placeholder,
        visibleLabel: widgetOptions.visibleLabel,
        ariaLabel: widgetOptions.ariaLabel,
        choices: widgetOptions.choices.map((choice) => choice.content),
    };
};

export default {
    name: "dropdown",
    displayName: "Drop down",
    defaultAlignment: "inline-block",
    accessible: true,
    widget: Dropdown,
    transform: optionsTransform,
    // TODO(LEMS-2656): remove TS suppression
    // @ts-expect-error: Type 'UserInput' is not assignable to type 'PerseusDropdownUserInput'.
    scorer: scoreDropdown,
} satisfies WidgetExports<typeof Dropdown>;
