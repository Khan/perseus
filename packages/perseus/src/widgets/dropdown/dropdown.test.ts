import {
    splitPerseusItem,
    type PerseusRenderer,
} from "@khanacademy/perseus-core";
import {scorePerseusItem} from "@khanacademy/perseus-score";
import {screen} from "@testing-library/react";
import {userEvent as userEventLib} from "@testing-library/user-event";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";
import {scorePerseusItemTesting} from "../../util/test-utils";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {basicDropdown} from "./dropdown.testdata";

import type {UserEvent} from "@testing-library/user-event";

describe("Dropdown widget", () => {
    let userEvent: UserEvent;
    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", async () => {
        // Arrange and Act
        const {container} = renderQuestion(basicDropdown);

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should snapshot when opened", async () => {
        // Arrange
        const {container} = renderQuestion(basicDropdown);

        // Act
        const dropdown = screen.getByRole("combobox");
        await userEvent.click(dropdown);

        // Assert
        expect(screen.getByText("less than or equal to")).toBeInTheDocument();
        expect(container).toMatchSnapshot("dropdown open");
    });

    it("should show placeholder text", async () => {
        // Arrange
        renderQuestion(basicDropdown);

        // Act
        const dropdown = screen.getByRole("combobox");

        // Assert
        // get asserts if it doesn't find a single matching element
        expect(dropdown).toHaveTextContent("greater/less than or equal to");
    });

    it("should be answerable correctly", async () => {
        // Arrange
        const {renderer} = renderQuestion(basicDropdown);

        // Act
        const dropdown = screen.getByRole("combobox");
        await userEvent.click(dropdown);
        await userEvent.click(screen.getByText("less than or equal to"));
        const score = scorePerseusItemTesting(
            basicDropdown,
            renderer.getUserInputMap(),
        );

        // Assert
        expect(dropdown).toHaveTextContent("less than or equal to");
        expect(score).toHaveBeenAnsweredCorrectly();
    });

    it("should be answerable incorrectly", async () => {
        // Arrange
        const {renderer} = renderQuestion(basicDropdown);

        // Act
        const dropdown = screen.getByRole("combobox");
        await userEvent.click(dropdown);
        await userEvent.click(screen.getByText("greater than or equal to"));
        const score = scorePerseusItemTesting(
            basicDropdown,
            renderer.getUserInputMap(),
        );

        // Assert
        expect(dropdown).toHaveTextContent("greater than or equal to");
        expect(score).toHaveBeenAnsweredIncorrectly();
    });

    it("should be invalid on first render", async () => {
        // Arrange
        const {renderer} = renderQuestion(basicDropdown);

        // Act
        const score = scorePerseusItemTesting(
            basicDropdown,
            renderer.getUserInputMap(),
        );

        // Assert
        expect(score).toHaveInvalidInput();
    });

    it("should be return true when focus() called", async () => {
        // Arrange
        const {renderer} = renderQuestion(basicDropdown);

        // Act
        const focused = renderer.focus();

        // Assert
        expect(focused).toBe(true);
        // TODO(LP-10797): we don't check that the document.activeElement is
        // actually set because the dropdown widget focuses a <div> (it's root
        // element), which is not actually focusable because it doesn't have a
        // tabindex.
    });

    it("has an ARIA label", async () => {
        // Arrange and Act
        renderQuestion(basicDropdown);

        // Assert
        expect(screen.getByLabelText("Select an answer")).toBeInTheDocument();
    });

    describe("interactive: full vs answerless", () => {
        beforeAll(() => {
            registerAllWidgetsForTesting();
        });

        let userEvent: UserEvent;
        beforeEach(() => {
            userEvent = userEventLib.setup({
                advanceTimers: jest.advanceTimersByTime,
            });

            jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
                testDependencies,
            );
        });

        function getAnswerfulItem(): PerseusRenderer {
            return {
                content: "[[â˜ƒ dropdown 1]]",
                images: {},
                widgets: {
                    "dropdown 1": {
                        type: "dropdown",
                        options: {
                            static: false,
                            placeholder: "Choose an answer",
                            choices: [
                                {
                                    content: "Correct",
                                    correct: true,
                                },
                                {
                                    content: "Incorrect",
                                    correct: false,
                                },
                            ],
                        },
                    },
                },
            };
        }

        function getAnswerlessItem(): PerseusRenderer {
            return splitPerseusItem(getAnswerfulItem());
        }

        test.each(["answerless", "answerful"])(
            "is interactive with widget options: %p",
            async (e) => {
                // Arrange
                const useAnswerless = e === "answerless";
                const renderItem = useAnswerless
                    ? getAnswerlessItem()
                    : getAnswerfulItem();

                // assert that splitting worked as expected
                if (useAnswerless) {
                    expect(
                        renderItem.widgets["dropdown 1"].options.choices[0]
                            .correct,
                    ).toBeUndefined();
                    expect(
                        renderItem.widgets["dropdown 1"].options.choices[1]
                            .correct,
                    ).toBeUndefined();
                }

                // Act
                const {renderer} = renderQuestion(renderItem);

                await userEvent.click(
                    screen.getByRole("combobox", {name: "Select an answer"}),
                );
                await userEvent.click(
                    screen.getByRole("option", {name: "Correct"}),
                );

                const userInput = renderer.getUserInputMap();
                const score = scorePerseusItem(
                    getAnswerfulItem(),
                    userInput,
                    "en",
                );

                // Assert
                expect(score).toHaveBeenAnsweredCorrectly();
            },
        );
    });
});
