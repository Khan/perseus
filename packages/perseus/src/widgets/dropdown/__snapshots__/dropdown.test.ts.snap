// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown widget should snapshot when opened: dropdown open 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        The total number of boxes the forklift can carry is 
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div
            class="default_xu2jcg"
          >
            <div
              class="default_xu2jcg-o_O-menuWrapper_wvrnr4"
            >
              <span
                aria-atomic="true"
                aria-live="polite"
                aria-relevant="additions text"
                class="srOnly_19bpjuy"
                data-testid="dropdown-live-region"
              >
                3 items
              </span>
              <button
                aria-controls=":r4:"
                aria-disabled="false"
                aria-expanded="true"
                aria-haspopup="listbox"
                aria-invalid="false"
                aria-label="Select an answer"
                class="button_vr44p2-o_O-shared_u51dsh-o_O-default_3ie67y"
                id=":r3:"
                role="combobox"
                type="button"
              >
                <span
                  class="text_f1191h-o_O-LabelMedium_1rew30o-o_O-text_16ujc3c"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        greater/less than or equal to
                      </div>
                    </div>
                  </div>
                </span>
                <span
                  aria-hidden="true"
                  class="svg_1q6jc65-o_O-icon_eocf2p-o_O-caret_ntkmq-o_O-inlineStyles_u7ovp6"
                />
              </button>
            </div>
          </div>
        </div>
         
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            60
          </span>
          <span />
        </span>
        .
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown widget should snapshot: initial render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        The total number of boxes the forklift can carry is 
        <div
          class="perseus-widget-container widget-nohighlight widget-inline-block"
        >
          <div
            class="default_xu2jcg"
          >
            <div
              class="default_xu2jcg-o_O-menuWrapper_wvrnr4"
            >
              <span
                aria-atomic="true"
                aria-live="polite"
                aria-relevant="additions text"
                class="srOnly_19bpjuy"
                data-testid="dropdown-live-region"
              />
              <button
                aria-controls=":r1:"
                aria-disabled="false"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-invalid="false"
                aria-label="Select an answer"
                class="button_vr44p2-o_O-shared_u51dsh-o_O-default_3ie67y"
                id=":r0:"
                role="combobox"
                type="button"
              >
                <span
                  class="text_f1191h-o_O-LabelMedium_1rew30o-o_O-text_16ujc3c"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        greater/less than or equal to
                      </div>
                    </div>
                  </div>
                </span>
                <span
                  aria-hidden="true"
                  class="svg_1q6jc65-o_O-icon_eocf2p-o_O-caret_ntkmq-o_O-inlineStyles_u7ovp6"
                />
              </button>
            </div>
          </div>
        </div>
         
        <span
          style="white-space: nowrap;"
        >
          <span />
          <span
            class="mock-TeX"
          >
            60
          </span>
          <span />
        </span>
        .
      </div>
    </div>
  </div>
</div>
`;
